docs/developers/napari_architecture.md

```diff
@@ -0,0 +1,164 @@
+# Napari architecture
+
+The napari codebase can thought to consist of three main components:
+
+* python models describing objects - these are able to operate separately from
+  the viewer and do not have any dependencies on user interface classes
+    * this code lives in `napari/components` (utility objects) and
+     `napari/layers` (objects that contain data)
+* qt classes that handle the interactive GUI aspect of the napari viewer
+    * the private qt code lives in `napari/_qt` and the smaller public qt
+      interface code lives in `napari/qt`
+* vispy classes that handles rendering
+    * the code for this is private and lives in `napari/_vispy`
+
+The separation of the python models from viewer GUI code allows:
+
+* the python model to be easily run headless without the viewer, for example
+  when performing batch analysis
+* analysis plugins are able to be developed without worrying about the GUI
+  aspect
+* napari has the option to move away from the rendering backend currently used
+* tests can easily be run headlessly
+
+## Python models and events
+
+Commonly, python models in napari are classes that store information about their
+state as an attribute and are the "source of ground truth". When these
+attributes are changed an "event" needs to be emitted such that relevant
+obserers of the model (such as other classes) can take the appropriate
+action. See [An Introduction to the Event Loop in napari](events/event_loop))
+for a little more background on events and the event-loop.
+
+One way this is achieved in napari is via getters and setters:
+
+```python
+from napari.utils.events import EmitterGroup
+
+class Weather:
```

Comment:
This example does a great job of explaining the Python models in napari and what boilerplate `EventedModel` absorbs! But it seems harder to reuse in the qt/vispy sections, which would be ideal. Maybe I'm being a boring pragmatist, but I think using a simplified version of one of the existing napari models could be a good idea. I think `Dims` might be a good target (but that's probably because I have dims/slicing on the mind) - that way you could show the slider in the Qt section and the canvas update in the vispy section.

Other alternatives could be `ViewerModel` and `LayerList`, though those might be a bit more complex.

Reply:
I had something like that but changed it to this. I like how the example shows how the same thing could be done with getter/setter and EventedModel base class. I guess I could take a Dims example but first show how it would be done with getter/setters.

Reply:
Yeah, it should be pretty easy to translate from the `Dims` model to the property getter/setter (I think that's what it used to be). It would probably be a good idea to use a simplified version of `Dims` (e.g. maybe with only `ndim`, `ndisplay`, and `current_step`). Sometimes I put `...` in doc code examples to indicate there is more in a class/function, but I'm not going to talk about it here because it's not relevant.

For Qt, we could talk about `QtDims` and `QtDimSliderWidget` event, commenting on how that connects to `current_step`. And for vispy, the effect of changing `ndisplay` might be a good one.